packages/landing/scripts/deploy.js

```diff
@@ -33,55 +34,68 @@ async function deploy() {
-
-            console.log('Uploading', filePath, `${(fileData.byteLength / 1024).toFixed(2)}Kb`, hash);
-            const res = await s3
-                .putObject({
-                    Bucket: s3BucketName,
-                    Key: filePath,
-                    Body: fileData,
-                    Metadata: { [HashKey]: hash },
-                    ContentType: mime.contentType(filePath),
-                    CacheControl: cacheControl,
-                })
-                .promise();
-
-            console.log('\tDone', res.ETag);
+    let changes = [];
+
+    async function upload(filePath, fileData, ContentType, CacheControl, hash, oldHash) {
+        if (oldHash != null) {
+            changes.push(filePath);
+        }
+
+        console.log('Uploading', filePath, `${(fileData.byteLength / 1024).toFixed(2)}Kb`, hash);
+        const res = await s3
+            .putObject({
+                Bucket: s3BucketName,
+                Key: filePath,
+                Body: fileData,
+                Metadata: { [HashKey]: hash },
+                ContentType,
+                CacheControl,
+            })
+            .promise();
+        console.log('\tDone', res.ETag);
+    }
+
+    // Upload all files from the dist directory, including any sub directories, to S3 bucket .
+    await recurseDirectory(DistDir, async (filePath, isDir) => {
+        if (isDir) return true;
+        const fileData = await fs.readFile(`${DistDir}/${filePath}`);
+        const hash = crypto.createHash('sha512').update(fileData).digest('base64');
+
+        const oldHash = await getHash(s3BucketName, filePath);
+        // Only upload files if they have changed
+        if (oldHash == hash) {
+            console.log('Skipped', filePath, `${(fileData.byteLength / 1024).toFixed(2)}Kb`, hash);
             return true;
-        });
-    };
+        }
+        const contentType = mime.contentType(extname(filePath));
 
-    await uploadDirectory();
+        // Don't set cache control for html files as they may have changed all other files are
+        // hashed and immutable
+        const cacheControl = filePath.endsWith('.html') ? undefined : 'public, max-age=604800, immutable';
 
-    if (hasChanges) await invalidateCache('/index.html', true);
+        await upload(filePath, fileData, contentType, cacheControl, hash, oldHash);
+
+        if (filePath.endsWith('/index.html')) {
+            // s3 does not handle subdirectory index.html so need to upload to dirname and
+            // dirname+'/' also
+            await upload(filePath.replace('/index.html', ''), fileData, contentType, cacheControl, hash, oldHash);
+            await upload(filePath.replace('/index.html', '/'), fileData, contentType, cacheControl, hash, oldHash);
+        }
+
+        return true;
+    });
+
+    for (const filePath of changes) {
+        await invalidateCache('/' + filePath, true);
```

Comment:
invalidations get quite pricey, we should batch this up into a single invalidation.

Reply:
Will now only invalidate at most `/index.html` and `/examples*`